python/django/security/injection/open-redirect.yaml

```diff
@@ -15,517 +16,37 @@ rules:
     - django
   languages: [python]
   severity: WARNING
-  patterns:
-  - pattern-inside: |
-      def $FUNC(...):
-        ...
-  - pattern-not-inside: |
-      def $FUNC(...):
-        ...
-        django.utils.http.is_safe_url(...)
-        ...
-  - pattern-not-inside: |
-      def $FUNC(...):
-        ...
-        if <... django.utils.http.is_safe_url(...) ...>:
+  pattern-sources:
+  - patterns:
+    - pattern-inside: |
+        def $FUNC(...):
           ...
+    - pattern-not-inside: |
+        def $FUNC(...):
+          ...
+          <... django.utils.http.is_safe_url(...) ...>
+          ...
+    - pattern-not-inside: |
+        def $FUNC(...):
+          ...
+          <... django.utils.http.url_has_allowed_host_and_scheme(...) ...>
+          ...
+    - pattern-not-inside: |
+        def $FUNC(...):
+          ...
+          if <... django.utils.http.is_safe_url(...) ...>:
+            ...
+    - pattern-not-inside: |
+        def $FUNC(...):
+          ...
+          if <... django.utils.http.url_has_allowed_host_and_scheme(...) ...>:
+            ...
+    - pattern-not: request.get_full_path(...)
```

Comment:
If we add `- pattern: request.$W` which is needed to catch some bugs in Django for example, then you need to write this as:
```
    - patterns:
      - pattern-either:
        - pattern: request.$W
        - pattern: request.$W.get(...)
        - pattern: request.$W(...)
        - pattern: request.$W[...]
      - metavariable-regex:
          metavariable: $W
          regex: (?!get_full_path)
```